content/Patterns for maintainable Rails applications. Part I - Ignacio Pérez.md

```diff
@@ -0,0 +1,111 @@
+
+https://gist.github.com/PerezIgnacio/e1babd9cab20f79131941c142b89b402
+
+Let’s assume that when the user signs up in our site, we also send a nested user_request with the filters. We receive this parameters in our controller and create both associated records in the database. In this case, the sign up controller may look something like this:
+
+https://gist.github.com/PerezIgnacio/b5e5212852fcbcff5b5e9c8e9bc74a2c
+
+This looks clean, user creation is simple because we have `accept_nested_attributes_for` in its model. However we are coupling our frontend to our database, while also forcing them to send the `user_request` param with `_attributes`. Also, if the user model and its associations gets more complex we may need to include additional code in the controller and, as we know, controllers should be as thin as possible. As an alternative, we can create a Form object, which could look similar to the following:
+
+https://gist.github.com/PerezIgnacio/da03bb1174565f63d392af570f78b592
+
+As you may have noticed, we need to add more code because nested attributes did things for us that we now have to do manually. However, this code is encapsulated in its own object, so it still is preferable over having code in controllers or models. Basically, the form object must be initialized with the params we receive in our controller and we will map to our models attributes. We define then a save method which is in charge of triggering the model persistence logic.
+
+We also include some validations from ActiveModel, checking the presence of some attributes in the form. All models can also be saved in a transaction, so it rollbacks in case any of the models created has an error. Additionally, we may add code that we want to execute after everything is created correctly, such as triggering a notification.
+
+The controller which uses the form object doesn’t change much. Here is how the form object could be used:
+
+https://gist.github.com/PerezIgnacio/d049e6bbb175de521d8456dcc7daf7e2
+
+In this case particularly, the form has all of its attributes explicitly defined, so there is no need to add strong parameters in our controllers, as we can ensure that there won’t be any additional attribute when persisting our models.
+
+Of course, this is a simple example, and the implementation will probably vary depending of the context where it’s used. Forms can easily get more complex, so in this cases it may be useful to combine this pattern with a Service Object, a pattern I will mention in the next part of this article.
+
+## Summary
+
+In this part of the article we went over some techniques used in Rails and the consequences of overusing them, while providing an alternative to each. We also reviewed principles that are necessary in order to understand the problems these solve. In the second part I will mention more of these techniques, and patterns we should use instead to improve our code maintainability.
```

Comment:
it'd be good to add a link to part 2 at the top and bottom of the page so that it's obvious to everyone. maybe even saying that this is a 2 posts series (?)